src/app/test-record/test-record.component.ts

```diff
@@ -15,29 +25,70 @@ import { TestType } from '@app/models/test.type';
   changeDetection: ChangeDetectionStrategy.OnPush
 })
 export class TestRecordComponent implements OnInit {
-  testResultObservable$: Observable<TestResultModel>;
-  testType: TestType;
-  testTypeNumber: string;
+  @Input() editState: VIEW_STATE;
+  @Input() preparers: Preparer[];
+  @Input() testResultObj: TestRecordTestType;
+  @Input() testStations: TestStation[];
+  @Input() testTypesApplicable: TestTypesApplicable;
+  @Output() submitTest = new EventEmitter<TestResultModel>();
+  @Output() switchState = new EventEmitter<VIEW_STATE>();
+  @Output() isFormDirty = new EventEmitter<boolean>();
+  testResultParentForm: FormGroup;
 
-  constructor(private _store: Store<IAppState>, private route: ActivatedRoute) {}
+  constructor(
+    private parent: FormGroupDirective,
+    protected fb: FormBuilder,
+    private dialog: MatDialog,
+    private testRecordMapper: TestRecordMapper
+  ) {}
 
-  ngOnInit() {
-    this.route.paramMap.subscribe((params) => {
-      this.testTypeNumber = params.get('id');
-    });
-    this.testResultObservable$ = this._store.pipe(select(selectSelectedVehicleTestResultModel));
-    this.testResultObservable$.subscribe((testResults) => {
-      let tType;
-      const tTypeNumber = this.testTypeNumber;
-      Object.keys(testResults).forEach(function(tRIndex) {
-        testResults[tRIndex].testTypes.some((res) => {
-          if (res.testNumber === tTypeNumber) {
-            tType = res;
-            return res;
-          }
-        });
+  ngOnInit(): void {
+    initAll();
+    this.switchState.emit(VIEW_STATE.VIEW_ONLY);
+    this.testResultParentForm = new FormGroup({ testType: new FormGroup({}) });
+    this.onFormChanges();
+  }
+
+  switchCurrentState(state: string) {
+    if (state === 'edit') {
+      this.switchState.emit(VIEW_STATE.EDIT);
+    } else {
+      const dialogRef = this.dialog.open(PreventLeavePageModalComponent, {
+        width: '45vw'
       });
-      this.testType = tType;
+
+      dialogRef.afterClosed().subscribe((result) => {
```

Comment:
this could be a separate function, it will make it easier to test, we could pass dialogRef as a dependency injection mock it and spy on it to check whether it is called or not